### ISSUE TITLE
Attempt to invoke virtual method on a null object reference - #570 reopened

### ISSUE THREAD

**Issue by FLATT-Dev** [don/cordova-plugin-ble-central#651]
I can reproduce this
The problem is in **Peripheral.java**, registerNotifyCallback()  method
`BluetoothGattService service = gatt.getService(serviceUUID);`
This returns null but no check is made and the service is passed to findNotifyCharacteristic() method where the code calls the service.getCharacteristics() method and because service is null  an exception is thrown.
I added some logging and a check to make sure service is not null - this is already used elsewhere in the code, and needs  to be added here probably as well

```
In registerNotifyCallback(...)
LOG.d(TAG,"registerNotifyCallback for service: "+serviceUUID+", charact: "+characteristicUUID);
BluetoothGattService service = gatt.getService(serviceUUID);
if (service == null) {
            callbackContext.error("Service " + serviceUUID + " not found.");
            LOG.d(TAG,"Service " + serviceUUID + " not found");
            return;
        }
```
However the curious thing is that the service does exist in the device I'm connecting to, and is found on another Android device (Galaxy Express)

**Blu model Vivo 5 Mini, Android v 6.0, kernel v 3.18.19**
12-13 10:46:46.585 11711 11726 D Peripheral: onConnectionStateChange CONNECTED
12-13 10:46:55.126 11711 11819 D Peripheral: requestMtu mtu=300
12-13 10:46:55.230 11711 11727 D Peripheral: mtu=300, status=0
12-13 10:47:14.792 11711 11819 D Peripheral: Queuing Command com.megster.cordova.ble.central.BLECommand@a4a3762
12-13 10:47:14.792 11711 11819 D Peripheral: Processing Commands
12-13 10:47:14.792 11711 11819 D Peripheral: Register Notify 00060001-f8ce-11e4-abf4-0002a5d5c51b
12-13 10:47:14.793 11711 11819 D Peripheral: registerNotifyCallback for service: 00060000-f8ce-11e4-abf4-0002a5d5c51b, charact: 00060001-f8ce-11e4-abf4-0002a5d5c51b
**12-13 10:47:14.793 11711 11819 D Peripheral: Service 00060000-f8ce-11e4-abf4-0002a5d5c51b not found**

**Samsung Galaxy Express Model Samsung-SM-J120A, Android v 6.0.1, kernel v 3.10.9-7662162**
12-13 10:55:52.401 26140 26203 D Peripheral: onConnectionStateChange CONNECTED
12-13 10:55:52.981 26140 26201 D Peripheral: requestMtu mtu=300
12-13 10:55:53.051 26140 26203 D Peripheral: mtu=300, status=0
12-13 10:55:53.491 26140 26201 D Peripheral: Queuing Command com.megster.cordova.ble.central.BLECommand@ceb074c
12-13 10:55:53.491 26140 26201 D Peripheral: Processing Commands
12-13 10:55:53.491 26140 26201 D Peripheral: Register Notify 00060001-f8ce-11e4-abf4-0002a5d5c51b
12-13 10:55:53.491 26140 26201 D Peripheral: registerNotifyCallback for service: 00060000-f8ce-11e4-abf4-0002a5d5c51b, charact: 00060001-f8ce-11e4-abf4-0002a5d5c51b
**The service is found and everything works as expected.**
Any ideas what can be done to fix this? I have full control of the device firmware so I can tweak things there if needed

Thanks
Andy

**Comment by FLATT-Dev** [comment]
OK I figured it out. The issue is that the device I connect to has 2 applications running. Each has its own set of services and characteristics which are cached by the OS  on connect and onServicesDiscovered callback gets the cached data.
If a different app is running next time I connect to the device, the cached services/characteristics from the previous connection are returned and that's what causes getService(uuid) call to fail.
I added refreshDeviceCache() call after connect, and now everything works.
Still might be a  good idea to check for null in service  =  gatt.getService(serviceUUID);  in registerNotifyCallback() - just as it is done in other places
Thanks
Andy
